packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+    public async attachOpHandler(
+        minSequenceNumber: number,
+        sequenceNumber: number,
+        term: number,
+        handler: IDeltaHandlerStrategy,
+        prefetchType: "cached" | "all" | "none" = "none",
+    ) {
+        this._virtualSequenceNumber = sequenceNumber;
+
+        this._sequenceNumberMap.set(sequenceNumber, sequenceNumber);
+        this._sequenceNumberMap.set(minSequenceNumber, minSequenceNumber);
+
+        return super.attachOpHandler(minSequenceNumber, sequenceNumber, term, handler, prefetchType);
+    }
+
+    protected connectHandler(connection: IConnectionDetails) {
+        // TODO: what else do we need to track here?
+        this._clientSequenceNumberMap.clear();
+        this._clientSequenceNumber = 0;
+        super.connectHandler(connection);
+    }
+
+    protected sendMessageToConnectionManager(messages: IDocumentMessage[]): void {
+        if (messages.length >= 2) {
+            this.connectionManager.sendMessages([{
+                type: GroupedBatchOpType,
+                clientSequenceNumber: ++this._clientSequenceNumber,
+                referenceSequenceNumber: messages[0].referenceSequenceNumber, // use oldest referenceSequenceNumber
+                contents: messages,
+            }]);
+        } else {
+            this.connectionManager.sendMessages(messages.map((it: IDocumentMessage) => {
+                this._clientSequenceNumberMap.set(++this._clientSequenceNumber, it.clientSequenceNumber);
+                return {...it, clientSequenceNumber: this._clientSequenceNumber};
+            }));
+        }
+    }
+
+    protected processMessage(message: ISequencedDocumentMessage, startTime: number): void {
+        if (message.type === GroupedBatchOpType) {
+            const messages = message.contents as IDocumentMessage[];
+            let groupSequenceNumber = this.virtualizeSequenceNumber(message.sequenceNumber) - messages.length + 1;
+
+            for (const subMessage of messages) {
+                super.processMessage({
+                    ...message, // This will override the property difference between ISequencedDocumentMessage and IDocumentMessage
+                    ...subMessage,
+                    sequenceNumber: groupSequenceNumber++,
+                    clientSequenceNumber: subMessage.clientSequenceNumber, // This number is already virtualized
+                    referenceSequenceNumber: this.virtualizeSequenceNumber(subMessage.referenceSequenceNumber),
+                    minimumSequenceNumber: this.virtualizeSequenceNumber(message.minimumSequenceNumber),
+                }, startTime);
+            }
+        } else {
+            super.processMessage({
+                ...message,
+                sequenceNumber: this.virtualizeSequenceNumber(message.sequenceNumber),
+                clientSequenceNumber: this.connectionManager.clientId === message.clientId ? this.virtualizeClientSequenceNumber(message.clientSequenceNumber) : message.clientSequenceNumber,
```

Comment:
This does not look right. Tracking code, this.connectionManager.clientId is current connection ID. But we could get our own ops (acks) from previous connection while already on new connection. This would be challenging to do right at this layer, as only Container (and ConnectionStateHandler) know about that extra layer of "virtualization" of clientSequence numbers.

Which actually raises an interesting question: there is no abstraction here, i.e. these numbers leak to other layers. We attempt to patch only them for our own ops. But there could be assumptions somewhere in the code that uses clientSequenceNumber for any op to do some kind of sorting or something else. So we would either need to keep a map for all clients (which is almost impossible, as we lack info when we join for the first time), OR we will need to store that virtualized nubmers on ops themselves, OR we need to get rid of clientSequenceNumber on the whole interface (ISequencedDocumentMessage)

Reply:
For prototype, you can start with simple - add it to payload (content), such that all clients have protocol and runtime clientSequenceNumber (for first message in the group). Ugly, but will do the work